include/llbuild/Core/BuildEngine.h

```diff
@@ -34,8 +34,39 @@
 namespace llbuild {
 namespace core {
 
-// FIXME: Need to abstract KeyType;
-typedef std::string KeyType;
+struct KeyType {
+  std::string function;
+  std::vector<uint8_t> args;
+
+  bool operator==(const KeyType& rhs) const {
+    return function == rhs.function && args == rhs.args;
+  }
+
+  bool operator<(const KeyType& rhs) const {
+    return function <= rhs.function && args < rhs.args;
+  }
+
+  std::string describe() const {
+    // FIXME: encode args
+    return function;
+  }
+
+  // FIXME: temporary compatibility methods
+  KeyType() { }
+  KeyType(const std::string& legacyKey) : function(legacyKey) { }
+  KeyType(llvm::StringRef legacyRef) : function(legacyRef) { }
+  KeyType(const char* str) : function(str) { }
+  KeyType(const unsigned char* const start, const unsigned char* const end) : function(start, end) { }
+  KeyType(const char* data, size_t length) : function(data, length) { }
+
+  const char* data() const { return function.data(); }
+  const char* c_str() const { return function.c_str(); }
+  size_t size() const { return function.size(); }
+  size_t length() const { return function.length(); }
```

Comment:
Do we need all these? Are these also just shims we can eventually eliminate?

If so, while not particularly important, it would be cleaner to refactor KeyType to its own type with just a single string inside of it, then refactor the clients, then do the key + args refactor.